presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/JoinGraphNode.java

```diff
@@ -0,0 +1,139 @@
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.PlanNodeId;
+import com.facebook.presto.sql.planner.plan.PlanVisitor;
+import com.facebook.presto.sql.tree.ComparisonExpression;
+import com.facebook.presto.sql.tree.Expression;
+import com.google.common.collect.ImmutableList;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.facebook.presto.sql.ExpressionUtils.and;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.INNER;
+import static com.facebook.presto.sql.tree.BooleanLiteral.TRUE_LITERAL;
+import static com.facebook.presto.sql.tree.ComparisonExpressionType.EQUAL;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This class represents a set of inner joins that can be executed in any order.
+ * The nodes of the join graph are the sources to be joined. The edges are
+ * filters. If there is no filter between a pair
+ * of sources, then a possible cross join is implied
+ */
+public class JoinGraphNode
+        extends PlanNode
+{
+    private final List<PlanNode> sources;
+    private final Expression filter;
+
+    public JoinGraphNode(PlanNodeId id, List<PlanNode> sources, Expression filter)
+    {
+        super(id);
+
+        requireNonNull(sources, "sources is null");
+        requireNonNull(filter, "filters is null");
+
+        this.sources = ImmutableList.copyOf(sources);
+        this.filter = filter;
+    }
+
+    public Expression getFilter()
+    {
+        return filter;
+    }
+
+    @Override
+    public <C, R> R accept(PlanVisitor<C, R> visitor, C context)
+    {
+        return visitor.visitJoinGraph(this, context);
+    }
+
+    @Override
+    public List<PlanNode> getSources()
+    {
+        return sources;
+    }
+
+    @Override
+    public List<Symbol> getOutputSymbols()
+    {
+        return sources.stream()
+                .flatMap(source -> source.getOutputSymbols().stream())
+                .collect(toImmutableList());
+    }
+
+    @Override
+    public PlanNode replaceChildren(List<PlanNode> newChildren)
+    {
+        return new JoinGraphNode(getId(), newChildren, filter);
+    }
+
+    //TODO: provide more efficient key implementation
+    public String getKey()
+    {
+        List<String> sourceIds = sources.stream()
+                .map(source -> source.getId().toString())
+                .collect(toImmutableList());
+        return sourceIds.toString() + filter.toString();
+    }
+
+    public static class JoinGraphNodeBuilder
+    {
+        private final List<PlanNode> sources = new ArrayList<>();
+        private final List<Expression> filters = new ArrayList<>();
+
+        // WARNING: this does not maintain the output symbols of the input joins.
+        // a project may need to be added above the JoinGraphNode to ensure correctness.
```

Comment:
Is this the responsibility of the called of this class to add that Project node? Is it possible to add that node in this class instead?